File: 1005-go-sensitive-word-filter/pkg/filter/trie.go
package filter

func newTrieNode() *TrieNode {
	return &TrieNode{
		children: make(map[rune]*TrieNode),
	}
}

func (t *TrieNode) insert(word string) {
	current := t
	for _, ch := range word {
		if _, exists := current.children[ch]; !exists {
			current.children[ch] = newTrieNode()
		}
		current = current.children[ch]
	}
	current.isEnd = true
	current.word = word
}

func (t *TrieNode) remove(word string) {
	var removeHelper func(node *TrieNode, word string, index int) bool
	removeHelper = func(node *TrieNode, word string, index int) bool {
		if index == len([]rune(word)) {
			if !node.isEnd {
				return false
			}
			node.isEnd = false
			node.word = ""
			return len(node.children) == 0
		}
		ch := []rune(word)[index]
		child, exists := node.children[ch]
		if !exists {
			return false
		}
		shouldRemove := removeHelper(child, word, index+1)
		if shouldRemove {
			delete(node.children, ch)
			return len(node.children) == 0 && !node.isEnd
		}
		return false
	}
	removeHelper(t, word, 0)
}

func (t *TrieNode) contains(word string) bool {
	current := t
	for _, ch := range word {
		child, exists := current.children[ch]
		if !exists {
			return false
		}
		current = child
	}
	return current.isEnd
}

func (t *TrieNode) listWords() []string {
	var words []string
	var collect func(node *TrieNode)
	collect = func(node *TrieNode) {
		if node.isEnd {
			words = append(words, node.word)
		}
		for _, child := range node.children {
			collect(child)
		}
	}
	collect(t)
	return words
}

File: 1659-go-react--OCR--/src/utils/trie.ts
class TrieNode {
  children: Map<string, TrieNode> = new Map();
  isEnd: boolean = false;
}

export class Trie {
  private root: TrieNode = new TrieNode();

  insert(word: string): void {
    let node = this.root;
    for (const char of word) {
      if (!node.children.has(char)) {
        node.children.set(char, new TrieNode());
      }
      node = node.children.get(char)!;
    }
    node.isEnd = true;
  }

  search(text: string): { word: string; start: number; end: number }[] {
    const results: { word: string; start: number; end: number }[] = [];
    const n = text.length;

    for (let i = 0; i < n; i++) {
      let node = this.root;
      let j = i;
      while (j < n && node.children.has(text[j])) {
        node = node.children.get(text[j])!;
        j++;
        if (node.isEnd) {
          results.push({
            word: text.substring(i, j),
            start: i,
            end: j
          });
        }
      }
    }

    return results;
  }
}
